Modern Toolbar / Drawer / Account Switcher

AndyScherzinger opened the pull request:
Use the new toolbar and support lib drawer implementation - Google's new, to be used implementation of the top bar instead of the action bar which can be considered legacy. Besides that the new implementation also sports an account switcher and avatars (images and default generated server-alike color circles).

**Motivation**
The actionbar is simply a bar which can have a menu, search widget and in newer AppCompat versions also a drawer implementation but isn't as flexible as the toolbar.

**Reasons and benefits**
The toolbar can be customized on many ways like but not limited to:
  * adding any kind of views to it, like the progressbar which would be nice especially when 
    * we integrate the multi select branch (no dark blue line below the then grey bar)
    * no dark blue line below the bar when the drawer is opened
  * animate it, like scrolling it away on a list/recycler view which seems nice on the file browsing views
  * the material drawer can be full height, like mentioned in #1210 because this is only possible for the toolbar not the action bar (it can be done with hacks though)

Material Design spec for side nav drawer: 
http://www.google.com/design/spec/layout/structure.html#structure-side-nav

przybylski commented on res/layout/account_action.xml:
```diff
@@ -0,0 +1,46 @@
+<!--
+  ownCloud Android client application
+
+  Copyright (C) 2016 ownCloud Inc.
```
RLY? ;]

AndyScherzinger replied:
Yes since this branch has been created in the owncloud repo first. So not sure how do deal with this?! @schiessle Any input is highly appreciated :)

AndyScherzinger replied:
added Copyright (C) 2016 Nextcloud

przybylski commented on res/layout/account_action.xml:
```diff
@@ -0,0 +1,46 @@
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_gravity="center_vertical"
+        android:layout_marginLeft="20dp"
```
constant

przybylski commented on res/layout/account_action.xml:
```diff
@@ -0,0 +1,46 @@
+        android:layout_weight="1"
+        android:gravity="center_vertical"
+        android:paddingRight="@dimen/standard_padding"
+        android:text="@string/placeholder_filename"
```
filename in account layout ?

AndyScherzinger replied:
yeah, just needed a placeholder so the rendering in Android Studio shows a value... found this one.... just used it ;) But yes, I should probably create a separate one.

przybylski commented on res/layout/accounts_layout.xml:
```diff
@@ -1,8 +1,8 @@
   ownCloud Android client application
 
-  Copyright (C) 2015 ownCloud Inc.
+  Copyright (C) 2016 ownCloud Inc.
```
maybe you want to add nexcloud copyright ?

AndyScherzinger replied:
again asking for your input @schiessle --> I wrote nearly all of the code in the ownCloud repo (like this file), then we forked and now I am finishing up the code. So what would be the best strategy as to decide when to add which copyright statement?

AndyScherzinger replied:
Copyright (C) 2016 Nextcloud

przybylski commented on src/com/owncloud/android/authentication/AccountUtils.java:
```diff
@@ -107,6 +107,20 @@ public static boolean exists(Account account, Context context) {
+    public static String getUsernameOfAccount(String accountName) {
+        if (accountName != null) {
+            return accountName.substring(0, accountName.lastIndexOf("@"));
+        } else {
```
I don't see a point of this else

przybylski commented on src/com/owncloud/android/authentication/AccountUtils.java:
```diff
@@ -107,6 +107,20 @@ public static boolean exists(Account account, Context context) {
+     * @param accountName the account name
+     * @return the user's name
+     */
+    public static String getUsernameOfAccount(String accountName) {
```
Wouldn't getAccountUsername be a better name?

przybylski commented on src/com/owncloud/android/datamodel/ThumbnailsCacheManager.java:
```diff
@@ -361,7 +360,223 @@ private Bitmap doFileInBackground() {
+    public static class AvatarGenerationTask extends AsyncTask<Object, Void, Bitmap> {
+        private final WeakReference<ImageView> mImageViewReference;
+        private final WeakReference<MenuItem> mMenuItemReference;
+        private static Account mAccount;
```
Why this is static?

AndyScherzinger replied:
pinging @tobiasKaminsky since you wrote the caching part for the avatars

tobiasKaminsky replied:
mAccount? I think it must be static as it is used in doInBackground and is not allowed to be changed

przybylski replied:
But mAccount is a part of async task which is alive throughout the entire life of AvatarGenerationTask. And so mAccount will not be changed. And you are assigning that only in constructor. If two AsyncTask's would be fired bad thing will happen. In the middle of the work someone can change account which will lead to hard to reconstruct bugs.

tobiasKaminsky replied:
So what is your suggestion? I am confused ;)

przybylski replied:
to make is non static member of the class

AndyScherzinger replied:
DONE, see https://github.com/nextcloud/android/pull/44/commits/54588809526e8437e78b369932f96a2a9bfb827e I also removed the storage manager @tobiasKaminsky since it isn't used.

AndyScherzinger commented:
Just FYI all comments made by @przybylski have been incorporated except https://github.com/nextcloud/android/pull/44#discussion_r67949040 which is still to be clarified first.

AndyScherzinger commented:
Can anybody test this, especially the account switching on a 4.1 or 4.2 device? Not an emulator!

tobiasKaminsky commented on src/com/owncloud/android/datamodel/ThumbnailsCacheManager.java:
```diff
@@ -361,7 +360,223 @@ private Bitmap doFileInBackground() {
+            return avatar;
+        }
+
+        private Bitmap handlePNG(Bitmap bitmap, int px){
```
not used?

AndyScherzinger replied:
Doesn't seem like, this is your own code though... So yes we do not handle PNG (so transparent background turn black, just tested it). and we also do not use doInBackround().  I just used your cache/task implementation from your former branch at the time and unfortunately lack any knwoledge about the image caching, async task, etc. As for the PNG issue I haven't found a way yet to find out that the image is a png. Any help, pointer are appriciated :)

tobiasKaminsky replied:
handlePNG is twice in ThumnailsCacheManager in:
AvatarGenerationTask
ThumbnailGenerationTask

Can avatars be png and therefore have to handled special?

AndyScherzinger replied:
Yes they can, I tested it but haven't found a way to detect png so the transparency turns black....  :(

AndyScherzinger replied:
handlePNG is now centralized 60f3a5a7ae37f2cb4d268e23d1afffad70b6c008

tobiasKaminsky commented on src/com/owncloud/android/datamodel/ThumbnailsCacheManager.java:
```diff
@@ -361,7 +360,223 @@ private Bitmap doFileInBackground() {
+            return resultBitmap;
+        }
+
+        private Bitmap doFileInBackground() {
```
not used?

AndyScherzinger replied:
As in my other comment. Doesn't seem like, this is your own code though... So yes it is unused but I can't tell you as to why that is. Any help, pointers, commits are appreciated :)

tobiasKaminsky commented on src/com/owncloud/android/ui/TextDrawable.java:
```diff
@@ -1,52 +1,126 @@
-        paint.setAntiAlias(true);
-        paint.setFakeBoldText(true);
+    /**
+     * creates an avatar in form of  a TextDrawable with the first letter of the account name in a circle with the
```
double space

AndyScherzinger replied:
DONE, see https://github.com/nextcloud/android/pull/44/commits/af1e0a8bf6f624f18e13b7bf425e086da90074a8

tobiasKaminsky commented on src/com/owncloud/android/ui/activity/BaseActivity.java:
```diff
@@ -0,0 +1,250 @@
+import com.owncloud.android.lib.resources.status.OCCapability;
+
+/**
+ * Base ctivity with common behaviour for activities dealing with ownCloud {@link Account}s .
```
missing "a"ctivity

AndyScherzinger replied:
DONE, see https://github.com/nextcloud/android/pull/44/commits/af1e0a8bf6f624f18e13b7bf425e086da90074a8

tobiasKaminsky commented on src/com/owncloud/android/ui/activity/BaseActivity.java:
```diff
@@ -0,0 +1,250 @@
+
+    /**
+     * Sets and validates the ownCloud {@link Account} associated to the Activity.
+     * <p/>
```
remove html tags

AndyScherzinger replied:
DONE, see https://github.com/nextcloud/android/pull/44/commits/af1e0a8bf6f624f18e13b7bf425e086da90074a8

tobiasKaminsky commented on src/com/owncloud/android/ui/activity/BaseActivity.java:
```diff
@@ -0,0 +1,250 @@
+     * <p/>
+     * If no valid ownCloud {@link Account} exists, the the user is requested
+     * to create a new ownCloud {@link Account}.
+     * <p/>
```
aalso here

AndyScherzinger replied:
DONE, see https://github.com/nextcloud/android/pull/44/commits/af1e0a8bf6f624f18e13b7bf425e086da90074a8

tobiasKaminsky commented on src/com/owncloud/android/ui/activity/ToolbarActivity.java:
```diff
@@ -0,0 +1,125 @@
+        View actionBarTitleView = getWindow().getDecorView().findViewById(
+                getResources().getIdentifier("action_bar_title", "id", "android")
+        );
+        // TODO remove legacy code
```
can this be removed?

AndyScherzinger replied:
DONE, see https://github.com/nextcloud/android/pull/44/commits/af1e0a8bf6f624f18e13b7bf425e086da90074a8

tobiasKaminsky commented on src/com/owncloud/android/utils/BitmapUtils.java:
```diff
@@ -266,5 +274,41 @@ public static boolean isImage(File file) {
+        String username = AccountUtils.getAccountUsername(accountName);
+        byte[] seed = username.getBytes("UTF-8");
+        MessageDigest md = MessageDigest.getInstance("MD5");
+//      Integer seedMd5Int = Math.abs(new String(Hex.encodeHex(seedMd5)).hashCode());
```
please remove

AndyScherzinger replied:
DONE, see https://github.com/nextcloud/android/pull/44/commits/af1e0a8bf6f624f18e13b7bf425e086da90074a8

tobiasKaminsky commented on src/com/owncloud/android/utils/DisplayUtils.java:
```diff
@@ -278,4 +287,52 @@ public static void colorSnackbar(Context context, Snackbar snackbar) {
+     * @param resources      reference for density information
+     * @param storageManager reference for caching purposes
+     */
+    public static void setAvatar(Account account, ImageView userIcon, float avatarRadius, Resources resources,
```
same code as in DrawerActivity.setAvatar?
At least the thumbnail caching and generation is the same.
Maybe simplify this?

AndyScherzinger replied:
Any hint on how to achieve this? I duplicated it since this is asynchronous and the other one isn't and I did get how I could centralize the code :(
I appreciate any input and hint you can give me :)

tobiasKaminsky replied:
Nope ;)
If if is not doable in an easy way it is fine for me.
cc @przybylski

AndyScherzinger replied:
I don't see how this could be simplified. Both methods implement the same flow but one is dealing with menu items and the other one with image views so besides the flow most other method calls differ completely. So I vote for leave it like this since the drawer implementation is specific to the menu item implementation and isn't used anywhere else while the other one could potentially be used on different screens.

tobiasKaminsky commented:
Code review done.
really really a great job!!!!
Unfortunately I do not have a 4.x device.

AndyScherzinger commented:
@tobiasKaminsky @przybylski I fixed the review comments and Bartosz's Fix for the avatar handling has been merged and tested on my device. I also added a refresh when the account has changed. So I think the PR is ready to go onto master. What do you think?

We could then finish up for a Play Store Beta release if that is fine with you :)

przybylski commented on src/com/owncloud/android/datamodel/ThumbnailsCacheManager.java:
```diff
@@ -323,21 +325,179 @@ private Bitmap doOCFileInBackground() {
+
+    }
+
+    public static class AvatarGenerationTask extends AsyncTask<Object, Void, Bitmap> {
```
You are always calling `execute` with String as a parameter, so it is a better idea to inherit from `AsyncTask<String, Void, Bitmap>` and avoid `instanceof` check for String

przybylski replied:
Unless I'm missing something

AndyScherzinger replied:
Nice catch! Fixed in https://github.com/nextcloud/android/pull/44/commits/1fab45537d5d4365c0898da11d9e9c729c4add3e

przybylski commented:
👍

tobiasKaminsky commented:
:+1:

AndyScherzinger commented:
:+1:
